File: tools/typescript/src/shared/stripe-client.ts
import type {Context} from './configuration';
import {StripeMcpClient, McpTool} from './mcp-client';
import {AsyncInitializer} from './async-initializer';

/**
 * Unified client for Stripe operations via MCP.
 *
 * All tool operations are executed via MCP (mcp.stripe.com).
 *
 * @example
 * const client = new StripeClient('rk_test_...', context);
 * await client.initialize();
 * const tools = client.getRemoteTools();
 * const result = await client.run('create_customer', { email: 'test@example.com' });
 * await client.close();
 */
class StripeClient {
  context: Context;

  private mcpClient: StripeMcpClient;
  private initializer = new AsyncInitializer();

  constructor(secretKey: string, context?: Context) {
    this.context = context || {};

    // MCP client for all tool operations
    this.mcpClient = new StripeMcpClient({
      secretKey,
      context: {
        account: context?.account,
        customer: context?.customer,
      },
      mode: context?.mode,
    });
  }

  /**
   * Async initialization - connects to MCP server.
   * Must be called before using tools via run().
   */
  initialize(): Promise<void> {
    return this.initializer.initialize(() => this.mcpClient.connect());
  }

  /**
   * Check if the client has been initialized.
   */
  isInitialized(): boolean {
    return this.initializer.isInitialized;
  }

  /**
   * Get tools from MCP server (after initialization).
   * Returns tool definitions with JSON Schema input schemas.
   */
  getRemoteTools(): McpTool[] {
    if (!this.initializer.isInitialized) {
      throw new Error(
        'StripeClient not initialized. Call initialize() before accessing tools.'
      );
    }
    return this.mcpClient.getTools();
  }

  /**
   * Execute a tool via MCP.
   * @param method - The tool method name (e.g., 'create_customer')
   * @param arg - The tool arguments
   * @param options - Optional per-call overrides (e.g., customer)
   * @returns JSON string result
   */
  run(
    method: string,
    arg: Record<string, unknown>,
    options?: {customer?: string}
  ): Promise<string> {
    if (!this.initializer.isInitialized) {
      throw new Error(
        'StripeClient not initialized. Call initialize() before running tools.'
      );
    }

    return this.mcpClient.callTool(method, arg, options);
  }

  /**
   * Close the MCP connection and clean up resources.
   * Safe to call multiple times (idempotent).
   */
  async close(): Promise<void> {
    if (!this.initializer.isInitialized) {
      return; // Already closed or never initialized
    }

    await this.mcpClient.disconnect();
    this.initializer.reset();
  }
}

export default StripeClient;

File: tools/python/stripe_agent_toolkit/shared/stripe_client.py
"""Unified client for Stripe operations via MCP."""

from typing import Optional, Dict, Any, List

from .mcp_client import StripeMcpClient, McpTool
from .async_initializer import AsyncInitializer
from ..configuration import Context


class StripeClient:
    """
    Unified client for Stripe operations via MCP.

    All tool operations are executed via MCP (mcp.stripe.com).

    Example:
        client = StripeClient('rk_test_...', context)
        await client.initialize()
        tools = client.get_remote_tools()
        result = await client.run('create_customer', {'email': 'test@example.com'})
        await client.close()
    """

    def __init__(
        self,
        secret_key: str,
        context: Optional[Context] = None
    ):
        self._context = context or {}
        self._initializer = AsyncInitializer()

        # MCP client for all tool operations
        self._mcp_client = StripeMcpClient({
            "secret_key": secret_key,
            "account": self._context.get("account"),
            "customer": self._context.get("customer"),
            "mode": self._context.get("mode"),
        })

    async def initialize(self) -> None:
        """Async initialization - connects to MCP server."""
        await self._initializer.initialize(self._mcp_client.connect)

    @property
    def is_initialized(self) -> bool:
        """Check if client is initialized."""
        return self._initializer.is_initialized

    def get_remote_tools(self) -> List[McpTool]:
        """Get tools from MCP server (after initialization)."""
        if not self._initializer.is_initialized:
            raise RuntimeError(
                "StripeClient not initialized. "
                "Call initialize() before accessing tools."
            )
        return self._mcp_client.get_tools()

    async def run(
        self,
        method: str,
        args: Dict[str, Any],
        customer: Optional[str] = None
    ) -> str:
        """
        Execute a tool via MCP.

        Args:
            method: Tool method name (e.g., 'create_customer')
            args: Tool arguments
            customer: Optional per-call customer override

        Returns:
            JSON string result
        """
        if not self._initializer.is_initialized:
            raise RuntimeError(
                "StripeClient not initialized. "
                "Call initialize() before running tools."
            )
        return await self._mcp_client.call_tool(method, args, customer)

    async def close(self) -> None:
        """Close MCP connection. Safe to call multiple times."""
        if not self._initializer.is_initialized:
            return

        await self._mcp_client.disconnect()
        self._initializer.reset()
